src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,376 @@
+% Concurrency
+
+Concurrency and parallelism are incredibly important topics in computer
+science, and are also a hot topic in industry today. Computers are gaining more
+and more cores, yet many programmers aren't prepared to fully utilize them.
+
+Rust's memory safety features also apply to its concurrency story too. Even
+concurrent Rust programs must be memory safe, having no data races. Rust's type
+system is up to the task, and gives you powerful ways to reason about
+concurrent code at compile time.
+
+Before we talk about the concurrency features that come with Rust, it's important
+to understand something: Rust is low-level enough that all of this is provided
+by the standard library, not by the language. This means that if you don't like
+some aspect of the way Rust handles concurrency, you can implement an alternative
+way of doing things. [mio](https://github.com/carllerche/mio) is a real-world
+example of this principle in action.
+
+## Background: `Send` and `Sync`
+
+Concurrency is difficult to reason about. In Rust, we have a strong, static
+type system to help us reason about our code. As such, Rust gives us two traits
+to help us make sense of code that can possibly be concurrent.
+
+### `Send`
+
+The first trait we're going to talk about is
+[`Send`](../std/marker/trait.Send.html). When a type `T` implements `Send`, it indicates
+to the compiler that something of this type is able to have ownership transferred
+safely between threads.
+
+This is important to enforce certain restrictions. For example, if we have a
+channel connecting two threads, we would want to be able to send some data
+down the channel and to the other thread. Therefore, we'd implement `Send` on
```

Comment:
I think the agency of "we'd implement" is subtly misleading. *Usually* you don't want to manually implement `Send`: if the compiler doesn't implement it automatically (i.e. a type contains something that is non-`Send`) there's a non-trivial risk that force-implementing it with `unsafe impl Send for ...` will be bad.

Maybe, just "Therefore, we'd ensure that `Send` is implemented"? not sure.

(And below, maybe "we would want to ensure that Send is not implemented".)

---

Brief overview to how `Send`/`Sync` interact with opt-in built-in traits: `Send` (resp. `Sync`) is automatically implemented for any type that contains only `Send` (resp. `Sync`) types. Users can override this default behaviour by force-implementing `Send` with an `unsafe impl`, e.g. when defining new threadsafe pointer types using raw pointers (although generally we try to recommend that something like `std::ptr::Unique` is used rather than a manual implementing the trait). For the other direction, users can use a negative impl to opt-out of a default `Send` implementation: `impl !Send for Foo {}`.